File: js/featurize.js
// Mirror of scripts/featurize.py. Turns a raw problem statement into named
// features, each carrying the span of text it came from.
//
// Where this fits: pattern-model.js feeds these features to the linear model
// to predict patterns, and the Analyze view uses the spans to highlight which
// parts of the problem drove each prediction.
//
// IMPORTANT — train/serve skew: this file and scripts/featurize.py must produce
// identical output for identical input. They are deliberately written as
// mechanical translations of each other; scripts/test_conformance.py checks
// them against shared fixtures. Change one, change both, then run that test.
// A silent divergence would make the browser explain a model that was never
// trained on these features.

export const REGIONS = ["statement", "example", "constraints"];
export const MAX_NGRAM = 3;

const HEADING_SOURCE =
  "^\\s*(input|output|constraints?|examples?|explanation|note|sample input|sample output)\\s*:?\\s*$";
const CONSTRAINT_HEADING_RE = /constraints?/i;
const EXAMPLE_HEADING_RE = /input|output|examples?|explanation|note|sample/i;
const TAG_RE = /<[^>]+>/g;
const PRE_RE = /<pre\b[^>]*>([\s\S]*?)<\/pre>/gi;
// Inline headings — a label with content following it on the same line.
const INLINE_EXAMPLE_SOURCE =
  "^[ \\t]*(examples?\\s*\\d*|inputs?|outputs?|explanation|sample\\s+(?:input|output)|note)\\s*:";

function headingRe() {
  return new RegExp(HEADING_SOURCE, "gim"); // fresh each time: lastIndex is stateful
}

// Entity decoding is done by hand rather than via a <textarea>, for two
// reasons: it must behave identically under Node so the conformance test can
// run there, and a DOM-based decoder would quietly resolve entities the Python
// side handles differently. Real statements use a small set — a survey of
// LeetCode HTML found only &quot;, &lt;, &nbsp; and &#39; — and numeric escapes
// are handled in full. &nbsp; maps to U+00A0 exactly as Python's
// html.unescape does, because strip() then collapses it to a space.
const NAMED_ENTITIES = {
  amp: "&", lt: "<", gt: ">", quot: '"', apos: "'", nbsp: " ",
  le: "≤", ge: "≥", ne: "≠", times: "×", minus: "−",
  hellip: "…", mdash: "—", ndash: "–", deg: "°",
  plusmn: "±", infin: "∞", rarr: "→", larr: "←",
  lsquo: "‘", rsquo: "’", ldquo: "“", rdquo: "”",
  sup2: "²", sup3: "³", frac12: "½",
};

function unescapeEntities(s) {
  if (!s.includes("&")) return s;
  return s.replace(/&(#x[0-9a-fA-F]+|#\d+|[a-zA-Z][a-zA-Z0-9]*);/g, (match, body) => {
    if (body[0] === "#") {
      const code = body[1] === "x" || body[1] === "X"
        ? parseInt(body.slice(2), 16)
        : parseInt(body.slice(1), 10);
      return Number.isFinite(code) && code >= 0 && code <= 0x10ffff
        ? String.fromCodePoint(code)
        : match;
    }
    const named = NAMED_ENTITIES[body];
    return named === undefined ? match : named;
  });
}

function strip(fragment) {
  let txt = fragment.replace(TAG_RE, " ");
  txt = unescapeEntities(txt);
  txt = txt.replace(/ /g, " ");
  txt = txt.replace(/[ \t]+/g, " ");
  return txt.replace(/\n{3,}/g, "\n\n");
}

function htmlToText(raw) {
  const out = [];
  const preSpans = [];
  let cursor = 0;
  let length = 0;
  PRE_RE.lastIndex = 0;
  let m;
  while ((m = PRE_RE.exec(raw)) !== null) {
    const before = strip(raw.slice(cursor, m.index));
    out.push(before);
    length += before.length;
    const inner = strip(m[1]);
    out.push(inner);
    preSpans.push([length, length + inner.length]);
    length += inner.length;
    cursor = m.index + m[0].length;
  }
  out.push(strip(raw.slice(cursor)));
  return [out.join(""), preSpans];
}

/** Raw statement (LeetCode HTML or plain text) -> display text + region spans.
 * Every span offset downstream refers to the returned text, so stripping
 * happens exactly once, here. */
export function toPlain(raw) {
  let text, preSpans;
  if (raw.includes("<") && raw.includes(">")) {
    [text, preSpans] = htmlToText(raw);
  } else {
    text = raw.replace(/\r\n/g, "\n");
    preSpans = [];
  }
  return { text, regions: regionsFor(text, preSpans) };
}

function nextHeadingOrEnd(text, fromPos) {
  const re = headingRe();
  re.lastIndex = fromPos;
  const m = re.exec(text);
  return m ? m.index : text.length;
}

function regionsFor(text, preSpans) {
  const spans = preSpans.map(([s, e]) => ["example", s, e]);

  const re = headingRe();
  let m;
  while ((m = re.exec(text)) !== null) {
    const end = nextHeadingOrEnd(text, m.index + m[0].length);
    if (CONSTRAINT_HEADING_RE.test(m[1])) spans.push(["constraints", m.index, end]);
    else if (EXAMPLE_HEADING_RE.test(m[1])) spans.push(["example", m.index, end]);
    if (re.lastIndex === m.index) re.lastIndex++; // zero-length match guard
  }

  // Headings that carry content on the same line, so the line-anchored pattern
  // above never sees them — "<strong>Constraints:</strong>" once tags are
  // stripped, and "Example 1:" / "Input: nums = [1,2]" in a plain-text paste.
  // See the matching note in featurize.py.
  const inlineExample = new RegExp(INLINE_EXAMPLE_SOURCE, "gim");
  while ((m = inlineExample.exec(text)) !== null) {
    spans.push(["example", m.index, nextHeadingOrEnd(text, m.index + m[0].length)]);
    if (inlineExample.lastIndex === m.index) inlineExample.lastIndex++;
  }
  const inlineConstraint = /constraints?\s*:/gi;
  while ((m = inlineConstraint.exec(text)) !== null) {
    spans.push(["constraints", m.index, nextHeadingOrEnd(text, m.index + m[0].length)]);
  }

  return flatten(spans, text.length);
}

const PRIORITY = { statement: 0, example: 1, constraints: 2 };

function flatten(spans, length) {
  if (!length) return [];
  const owner = new Array(length).fill("statement");
  for (const [region, start, end] of spans) {
    for (let i = Math.max(0, start); i < Math.min(length, end); i++) {
      if (PRIORITY[region] >= PRIORITY[owner[i]]) owner[i] = region;
    }
  }
  const out = [];
  let cur = owner[0];
  let start = 0;
  for (let i = 1; i < length; i++) {
    if (owner[i] !== cur) {
      out.push([cur, start, i]);
      cur = owner[i];
      start = i;
    }
  }
  out.push([cur, start, length]);
  return out;
}

export function regionAt(regions, pos) {
  for (const [region, start, end] of regions) {
    if (pos >= start && pos < end) return region;
  }
  return "statement";
}

// Mirrors _TOKEN_RE in featurize.py — see the note there. Subscripted names
// reach us as "starti" from HTML and "start_i" from a plain-text paste; both
// must tokenize the same way or a pasted problem loses the very features the
// model learned.
const TOKEN_RE = /[a-zA-Z]+(?:_[a-zA-Z0-9]+)*|\d+/g;
const STOPWORDS = new Set(`
a an the and or of to in is are be been was were it its this that these those
for with as at by on from you your we our they their he she his her i if then
else so such can could will would should may might must do does did done have
has had not no nor but there here what which who whom when where why how all
any both each few more most other some only own same than too very s t just
`.trim().split(/\s+/));

/** Lowercased tokens with positions. Integers collapse to a magnitude token so
 * "10^5" and "100000" don't each claim their own vocabulary entry. */
export function tokenize(text, regions) {
  const tokens = [];
  TOKEN_RE.lastIndex = 0;
  let m;
  while ((m = TOKEN_RE.exec(text)) !== null) {
    const raw = m[0];
    let tok;
    if (/^\d+$/.test(raw)) {
      tok = `<num${raw.length}>`;
    } else {
      tok = raw.toLowerCase().replace(/_/g, "");
      if (STOPWORDS.has(tok) || tok.length <= 1) continue;
    }
    tokens.push([tok, m.index, m.index + raw.length, regionAt(regions, m.index)]);
  }
  return tokens;
}

/** { feature: [[start, end, region], ...] } for 1..MAX_NGRAM grams. An n-gram
 * is only formed from tokens sharing a region, so a phrase never straddles the
 * boundary between the statement and a constraints list. */
export function ngramFeatures(tokens) {
  const feats = new Map();
  for (let n = 1; n <= MAX_NGRAM; n++) {
    for (let i = 0; i + n <= tokens.length; i++) {
      const window = tokens.slice(i, i + n);
      const region = window[0][3];
      if (window.some((t) => t[3] !== region)) continue;
      const name = "w:" + window.map((t) => t[0]).join(" ");
      if (!feats.has(name)) feats.set(name, []);
      feats.get(name).push([window[0][1], window[n - 1][2], region]);
    }
  }
  return feats;
}

// ---------- constraint parsing ----------

// Mirrors _NUM / _BOUND_RE in featurize.py. The number pattern deliberately
// forbids bare whitespace between digits: allowing it made "2000 \n 0 <= ..."
// parse as the single number 20000.
const NUM = "-?\\d+(?:\\.\\d+)?(?:\\s*(?:\\^|\\*\\*)\\s*\\d+|[eE]\\+?\\d+)?";
const BOUND_SOURCE =
  `(?:(?<lo>${NUM})\\s*(?:<=|≤|<)\\s*)?` +
  "(?<var>[A-Za-z_][A-Za-z0-9_]*(?:\\s*\\[\\s*[A-Za-z0-9_]+\\s*\\])?(?:\\s*\\.\\s*(?:length|size|len))?)" +
  `\\s*(?:<=|≤|<|==|=)\\s*(?<hi>${NUM})`;
const LEN_SUFFIX_RE = /\.\s*(length|size|len)\s*$/i;
const SIZE_WORD_RE = /(^|_)(n|m|k|q|len|length|size|count|total|num|rows?|cols?|columns?|nodes?|edges?|vertices)($|_)|num|count|len|size|total/i;

export const SIZE_BUCKETS = [[20, "le20"], [100, "le100"], [1000, "le1e3"],
  [10000, "le1e4"], [100000, "le1e5"], [1000000, "le1e6"]];
export const VALUE_BUCKETS = [[100, "le100"], [10000, "le1e4"], [1e9, "le1e9"]];

/** Size-like (how much input) vs value-like (how big an element). A size bound
 * drives the complexity class you can afford; a value bound hints at counting
 * sorts and bit tricks. An inner array's arity ("prerequisites[i].length == 2")
 * is neither, and is filed as a value so it can't masquerade as a tiny input. */
function boundKind(varName) {
  const hasIndex = varName.includes("[");
  const hasLen = LEN_SUFFIX_RE.test(varName);
  if (hasLen && !hasIndex) return "size";
  if (hasIndex) return "value";
  return SIZE_WORD_RE.test(varName) ? "size" : "value";
}

function parseNumber(raw) {
  const s = raw.replace(/\s+/g, "");
  let m = /^(-?\d+)(?:\^|\*\*)(\d+)$/.exec(s);
  if (m) {
    const exp = parseInt(m[2], 10);
    return exp <= 18 ? Math.pow(parseInt(m[1], 10), exp) : null;
  }
  m = /^(-?\d+(?:\.\d+)?)[eE]\+?(\d+)$/.exec(s);
  if (m) {
    const exp = parseInt(m[2], 10);
    return exp <= 18 ? Math.round(parseFloat(m[1]) * Math.pow(10, exp)) : null;
  }
  if (/^-?\d+$/.test(s)) return parseInt(s, 10);
  return null;
}

function bucket(value, table, overflow) {
  for (const [limit, name] of table) if (value <= limit) return name;
  return overflow;
}

/** Features from the constraints region only. */
export function constraintFeatures(text, regions) {
  const feats = new Map();
  let maxSize = 0;
  let maxValue = 0;
  let sawNegative = false;
  let sizeSpan = null;
  let valueSpan = null;

  for (const [region, start, end] of regions) {
    if (region !== "constraints") continue;
    const chunk = text.slice(start, end);
    const re = new RegExp(BOUND_SOURCE, "g");
    let m;
    while ((m = re.exec(chunk)) !== null) {
      const hi = parseNumber(m.groups.hi);
      if (hi === null || hi <= 0) {
        if (hi !== null && hi < 0) sawNegative = true;
        continue;
      }
      if (m.groups.lo != null) {
        const lo = parseNumber(m.groups.lo);
        if (lo !== null && lo < 0) sawNegative = true;
      }
      const span = [start + m.index, start + m.index + m[0].length, "constraints"];
      if (boundKind(m.groups.var) === "size") {
        if (hi > maxSize) { maxSize = hi; sizeSpan = span; }
      } else if (hi > maxValue) {
        maxValue = hi; valueSpan = span;
      }
    }
  }

  const add = (name, span) => {
    if (!feats.has(name)) feats.set(name, []);
    feats.get(name).push(span);
  };
  if (maxSize) add(`c:size:${bucket(maxSize, SIZE_BUCKETS, "gt1e6")}`, sizeSpan);
  if (maxValue) add(`c:value:${bucket(maxValue, VALUE_BUCKETS, "gt1e9")}`, valueSpan);
  if (sawNegative) add("c:has_negative", [0, 0, "constraints"]);
  if (!maxSize && !maxValue) add("c:none", [0, 0, "constraints"]);
  return feats;
}

/** Full pipeline. Returns { text, regions, features } where features maps a
 * feature name to the list of [start, end, region] spans it occurred at. */
export function featurize(raw) {
  const { text, regions } = toPlain(raw);
  const tokens = tokenize(text, regions);
  const features = ngramFeatures(tokens);
  for (const [name, spans] of constraintFeatures(text, regions)) {
    if (!features.has(name)) features.set(name, []);
    features.get(name).push(...spans);
  }
  return { text, regions, features };
}

File: scripts/featurize.py
#!/usr/bin/env python3
"""Turns a raw problem statement into named features, with the span of text
each feature came from.

Where this fits: used by train_model.py to build the training matrix, and
mirrored exactly by js/featurize.js at inference time in the browser.

IMPORTANT — train/serve skew: this module and js/featurize.js must
produce identical output for identical input. They are deliberately written
as mechanical translations of each other, and scripts/test_conformance.py
checks them against shared fixtures. Change one, change both, then run that
test; a silent divergence would make the browser's explanations describe a
model that was never trained.

Two design decisions worth stating:

1. Features carry spans. An n-gram isn't just "contiguous subarray" — it's
   that phrase at characters 48-67, inside the statement region. The
   explanation UI needs those offsets to highlight text, and a linear model
   makes each span's contribution exactly weight x count, so the highlight is
   the real reason rather than an approximation of it.

2. Constraint bounds are parsed, not scraped. An early prototype took the
   largest integer anywhere in the text, which happily read "40 tournaments"
   as an input bound and put only 4 of 347 problems in the "n is tiny" bucket.
   Bounds are now matched only inside the constraints region and bound to a
   variable name, then split into size-like bounds (how big is the input —
   drives complexity class) and value-like bounds (how big are the elements —
   hints at counting/bit tricks).
"""
import html
import re

# ---------- region segmentation ----------

REGIONS = ("statement", "example", "constraints")

# LeetCode wraps examples in <pre> and prefixes the constraints list with a
# bolded "Constraints:". Codeforces/CodeChef plain text uses bare headings.
_HEADING_RE = re.compile(
    r"^\s*(input|output|constraints?|examples?|explanation|note|sample input|sample output)\s*:?\s*$",
    re.I | re.M,
)
_CONSTRAINT_HEADING_RE = re.compile(r"constraints?", re.I)
_EXAMPLE_HEADING_RE = re.compile(r"input|output|examples?|explanation|note|sample", re.I)

_TAG_RE = re.compile(r"<[^>]+>")
_PRE_RE = re.compile(r"<pre\b[^>]*>(.*?)</pre>", re.I | re.S)

# Inline headings — a label with content following it on the same line.
_INLINE_EXAMPLE_RE = re.compile(
    r"^[ \t]*(examples?\s*\d*|inputs?|outputs?|explanation|sample\s+(?:input|output)|note)\s*:",
    re.I | re.M,
)
_INLINE_CONSTRAINT_RE = re.compile(r"constraints?\s*:", re.I)


def to_plain(raw):
    """Convert a raw statement (LeetCode HTML or plain text) into display text
    plus region spans. Returns (text, [(region, start, end), ...]).

    The returned text is what the UI renders and what every span offset refers
    to, so stripping happens once, here, and never again downstream."""
    if "<" in raw and ">" in raw:
        text, pre_spans = _html_to_text(raw)
    else:
        text, pre_spans = raw.replace("\r\n", "\n"), []
    return text, _regions_for(text, pre_spans)


def _html_to_text(raw):
    """Strip tags, decode entities, and remember where <pre> blocks landed."""
    out = []
    pre_spans = []
    cursor = 0
    for m in _PRE_RE.finditer(raw):
        out.append(_strip(raw[cursor:m.start()]))
        start = sum(len(s) for s in out)
        out.append(_strip(m.group(1)))
        pre_spans.append((start, sum(len(s) for s in out)))
        cursor = m.end()
    out.append(_strip(raw[cursor:]))
    return "".join(out), pre_spans


def _strip(fragment):
    txt = _TAG_RE.sub(" ", fragment)
    txt = html.unescape(txt)
    txt = txt.replace("\xa0", " ")
    txt = re.sub(r"[ \t]+", " ", txt)
    return re.sub(r"\n{3,}", "\n\n", txt)


def _regions_for(text, pre_spans):
    """Label every character as statement, example, or constraints.

    Constraints win over examples where they overlap: LeetCode puts the
    constraints list after the last <pre>, and the bounds inside it are the
    single most informative thing in the whole problem."""
    spans = [("example", s, e) for s, e in pre_spans]

    for m in _HEADING_RE.finditer(text):
        label = m.group(1)
        end = _next_heading_or_end(text, m.end())
        if _CONSTRAINT_HEADING_RE.search(label):
            spans.append(("constraints", m.start(), end))
        elif _EXAMPLE_HEADING_RE.search(label):
            spans.append(("example", m.start(), end))

    # Headings that carry content on the same line, so the line-anchored
    # pattern above never sees them. Both spellings occur constantly:
    # "<strong>Constraints:</strong>" collapses to inline text once tags are
    # stripped, and a plain-text paste writes "Example 1:" or
    # "Input: nums = [1,2]". Without these, a pasted problem put its worked
    # examples in the statement region and the per-region breakdown read
    # "example: 0.00" for every plain-text problem.
    for m in _INLINE_EXAMPLE_RE.finditer(text):
        spans.append(("example", m.start(), _next_heading_or_end(text, m.end())))
    for m in _INLINE_CONSTRAINT_RE.finditer(text):
        spans.append(("constraints", m.start(), _next_heading_or_end(text, m.end())))

    return _flatten(spans, len(text))


def _next_heading_or_end(text, from_pos):
    nxt = _HEADING_RE.search(text, from_pos)
    return nxt.start() if nxt else len(text)


def _flatten(spans, length):
    """Resolve overlaps into a non-overlapping cover, constraints highest
    priority, and fill the gaps with statement."""
    priority = {"statement": 0, "example": 1, "constraints": 2}
    owner = ["statement"] * length
    for region, start, end in spans:
        for i in range(max(0, start), min(length, end)):
            if priority[region] >= priority[owner[i]]:
                owner[i] = region
    out = []
    if not length:
        return out
    cur, start = owner[0], 0
    for i in range(1, length):
        if owner[i] != cur:
            out.append((cur, start, i))
            cur, start = owner[i], i
    out.append((cur, start, length))
    return out


def region_at(regions, pos):
    for region, start, end in regions:
        if start <= pos < end:
            return region
    return "statement"


# ---------- tokenization ----------

# Subscripted identifiers must survive tokenization as one token. The same
# variable reaches us two ways: LeetCode HTML renders "start<sub>i</sub>",
# which flattens to "starti", while a plain-text paste says "start_i". Splitting
# on the underscore produced "start" plus a single letter that the length filter
# then discarded, so a pasted problem lost exactly the features the model had
# learned from the HTML — merge-intervals text ranked 5th for intervals because
# of it. Matching the underscored form and stripping the underscores makes both
# spellings land on the same token.
_TOKEN_RE = re.compile(r"[a-zA-Z]+(?:_[a-zA-Z0-9]+)*|\d+")
# Identifier-ish noise that carries no pattern signal but bloats the vocab.
STOPWORDS = frozenset("""
a an the and or of to in is are be been was were it its this that these those
for with as at by on from you your we our they their he she his her i if then
else so such can could will would should may might must do does did done have
has had not no nor but there here what which who whom when where why how all
any both each few more most other some only own same than too very s t just
""".split())


def tokenize(text, regions):
    """Lowercased alphabetic tokens with positions; integers collapse to a
    magnitude token so "10^5" and "100000" don't each claim vocabulary."""
    tokens = []
    for m in _TOKEN_RE.finditer(text):
        raw = m.group(0)
        if raw.isdigit():
            tok = f"<num{len(raw)}>"
        else:
            tok = raw.lower().replace("_", "")
            if tok in STOPWORDS or len(tok) <= 1:
                continue
        tokens.append((tok, m.start(), m.end(), region_at(regions, m.start())))
    return tokens


MAX_NGRAM = 3


def ngram_features(tokens):
    """{feature: [(start, end, region), ...]} for 1..MAX_NGRAM grams.

    An n-gram is only formed from tokens sharing a region, so a phrase never
    straddles the boundary between the statement and a constraints list."""
    feats = {}
    for n in range(1, MAX_NGRAM + 1):
        for i in range(len(tokens) - n + 1):
            window = tokens[i:i + n]
            if len({t[3] for t in window}) != 1:
                continue
            name = "w:" + " ".join(t[0] for t in window)
            feats.setdefault(name, []).append((window[0][1], window[-1][2], window[0][3]))
    return feats


# ---------- constraint parsing ----------

# A number literal: 2000, 10^5, 10 ** 5, 1e9. Deliberately does NOT allow bare
# whitespace between digits — an earlier version did, and greedily read
# "2000 \n 0 <= ..." as the single number 20000.
_NUM = r"-?\d+(?:\.\d+)?(?:\s*(?:\^|\*\*)\s*\d+|[eE]\+?\d+)?"
# Matches "1 <= n <= 10^5", "n ≤ 100", "0 <= nums[i] <= 10^9", "n == 20".
_BOUND_RE = re.compile(
    rf"(?:(?P<lo>{_NUM})\s*(?:<=|≤|<)\s*)?"
    r"(?P<var>[A-Za-z_][A-Za-z0-9_]*(?:\s*\[\s*[A-Za-z0-9_]+\s*\])?(?:\s*\.\s*(?:length|size|len))?)"
    rf"\s*(?:<=|≤|<|==|=)\s*(?P<hi>{_NUM})"
)
_INDEX_RE = re.compile(r"\[")
_LEN_SUFFIX_RE = re.compile(r"\.\s*(length|size|len)\s*$", re.I)
# Names that denote how much input there is rather than how large a value is.
_SIZE_WORD_RE = re.compile(r"(^|_)(n|m|k|q|len|length|size|count|total|num|rows?|cols?|columns?|nodes?|edges?|vertices)($|_)|num|count|len|size|total", re.I)


def _bound_kind(var):
    """Size-like (how much input) vs value-like (how big an element).

    The distinction matters: a size bound drives the complexity class you can
    afford, while a value bound hints at counting sorts and bit tricks. An
    inner array's arity ("prerequisites[i].length == 2") is neither, and is
    filed as a value so it cannot masquerade as a tiny input size."""
    has_index = bool(_INDEX_RE.search(var))
    has_len = bool(_LEN_SUFFIX_RE.search(var))
    if has_len and not has_index:
        return "size"
    if has_index:
        return "value"
    return "size" if _SIZE_WORD_RE.search(var) else "value"

SIZE_BUCKETS = [(20, "le20"), (100, "le100"), (1000, "le1e3"), (10000, "le1e4"),
                (100000, "le1e5"), (1000000, "le1e6")]
VALUE_BUCKETS = [(100, "le100"), (10000, "le1e4"), (10 ** 9, "le1e9")]


def _parse_number(raw):
    """Read 10^5, 10**5, 1e5, or a plain integer. Returns None if implausible."""
    s = re.sub(r"\s+", "", raw)
    m = re.fullmatch(r"(-?\d+)(?:\^|\*\*)(\d+)", s)
    if m:
        base, exp = int(m.group(1)), int(m.group(2))
        return base ** exp if exp <= 18 else None
    m = re.fullmatch(r"(-?\d+(?:\.\d+)?)[eE]\+?(\d+)", s)
    if m:
        exp = int(m.group(2))
        return int(float(m.group(1)) * 10 ** exp) if exp <= 18 else None
    if re.fullmatch(r"-?\d+", s):
        return int(s)
    return None


def _bucket(value, table, overflow):
    for limit, name in table:
        if value <= limit:
            return name
    return overflow


def constraint_features(text, regions):
    """Features from the constraints region only, split into size-like and
    value-like bounds. Returns {feature: [(start, end, 'constraints')]}."""
    feats = {}
    max_size, max_value, saw_negative = 0, 0, False
    size_span, value_span = None, None

    for region, start, end in regions:
        if region != "constraints":
            continue
        chunk = text[start:end]
        for m in _BOUND_RE.finditer(chunk):
            hi = _parse_number(m.group("hi"))
            if hi is None or hi <= 0:
                if hi is not None and hi < 0:
                    saw_negative = True
                continue
            lo_raw = m.group("lo")
            if lo_raw is not None and (lo := _parse_number(lo_raw)) is not None and lo < 0:
                saw_negative = True
            span = (start + m.start(), start + m.end(), "constraints")
            if _bound_kind(m.group("var")) == "size":
                if hi > max_size:
                    max_size, size_span = hi, span
            elif hi > max_value:
                max_value, value_span = hi, span

    if max_size:
        feats.setdefault(f"c:size:{_bucket(max_size, SIZE_BUCKETS, 'gt1e6')}", []).append(size_span)
    if max_value:
        feats.setdefault(f"c:value:{_bucket(max_value, VALUE_BUCKETS, 'gt1e9')}", []).append(value_span)
    if saw_negative:
        feats.setdefault("c:has_negative", []).append((0, 0, "constraints"))
    if not max_size and not max_value:
        feats.setdefault("c:none", []).append((0, 0, "constraints"))
    return feats


# ---------- top level ----------

def featurize(raw):
    """Full pipeline. Returns (text, regions, {feature: [(start, end, region)]})."""
    text, regions = to_plain(raw)
    tokens = tokenize(text, regions)
    feats = ngram_features(tokens)
    for name, spans in constraint_features(text, regions).items():
        feats.setdefault(name, []).extend(spans)
    return text, regions, feats


def feature_counts(raw):
    """Just {feature: count} — what the training matrix needs."""
    _, _, feats = featurize(raw)
    return {name: len(spans) for name, spans in feats.items()}
